packages/contracts-bedrock/src/dispute/FaultDisputeGame2.sol

```diff
@@ -942,4 +951,35 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
             ? Hash.wrap(keccak256(abi.encode(_disputed, _disputedPos)))
             : Hash.wrap(keccak256(abi.encode(_starting, _startingPos, _disputed, _disputedPos)));
     }
+
+    function _traceAncestorV2(Position _position, uint256 _intervalDepth)
+        internal
+        pure
+        returns (Position ancestor_)
+    {
+        if (_position.depth() % _intervalDepth != 0 || _position.depth() < _intervalDepth) {
+            revert InvalidClaim();
+        }
+        ancestor_ = _position.traceAncestor();
+        // Position 1 is not the value submitted by the Actors.
+        if (1 == ancestor_.raw()) {
+            return ancestor_;
+        }
+        return _firstValidPosition(ancestor_, _intervalDepth);
+    }
+
+    /// @notice The position of the first claim submitted in the current location or subtree.
+    /// @param _position Current location
+    /// @param _intervalDepth The depth of the interval with each submission.
+    /// @return first_ First valid location
+    function _firstValidPosition(Position _position, uint256 _intervalDepth)
```

Comment:
`first` may be confusing since it can be vertical (depth) or horizontal.

I think we can use `validRightIndex` or `firstValidRightIndex`.